rider-fsharp/src/main/java/com/jetbrains/rider/plugins/fsharp/services/fsi/FsiHost.kt

```diff
@@ -1,22 +1,119 @@
 package com.jetbrains.rider.plugins.fsharp.services.fsi
 
+import com.intellij.execution.Executor
+import com.intellij.execution.configurations.GeneralCommandLine
+import com.intellij.execution.process.ProcessAdapter
+import com.intellij.execution.process.ProcessEvent
 import com.intellij.notification.Notification
 import com.intellij.notification.NotificationType
 import com.intellij.notification.Notifications
 import com.intellij.openapi.editor.Editor
 import com.intellij.openapi.project.Project
-import com.intellij.openapi.util.io.FileUtil
+import com.intellij.openapi.project.guessProjectDir
+import com.intellij.openapi.util.NlsSafe
+import com.intellij.openapi.vfs.VfsUtil
+import com.intellij.project.isDirectoryBased
 import com.intellij.psi.PsiFile
-import com.jetbrains.rdclient.util.idea.LifetimedProjectComponent
-import com.jetbrains.rider.projectView.solution
+import com.intellij.util.ui.UIUtil
 import com.jetbrains.rd.util.reactive.Property
 import com.jetbrains.rd.util.reactive.flowInto
+import com.jetbrains.rdclient.util.idea.LifetimedProjectComponent
 import com.jetbrains.rider.plugins.fsharp.*
+import com.jetbrains.rider.plugins.fsharp.services.fsi.consoleRunners.FsiConsoleRunnerBase
+import com.jetbrains.rider.plugins.fsharp.services.fsi.consoleRunners.FsiDefaultConsoleRunner
+import com.jetbrains.rider.plugins.fsharp.services.fsi.consoleRunners.FsiScriptProfileConsoleRunner
+import com.jetbrains.rider.plugins.fsharp.services.fsi.runScript.FSharpScriptConfiguration
+import com.jetbrains.rider.projectView.solution
+import com.jetbrains.rider.run.configurations.dotNetExe.DotNetExeConfigurationParameters
+import com.jetbrains.rider.run.createRunCommandLine
+import com.jetbrains.rider.runtime.DotNetExecutable
+import com.jetbrains.rider.runtime.DotNetRuntime
+import com.jetbrains.rider.runtime.RiderDotNetActiveRuntimeHost
+import com.jetbrains.rider.runtime.dotNetCore.DotNetCoreRuntimeType
+import com.jetbrains.rider.runtime.mono.MonoRuntime
+import com.jetbrains.rider.runtime.mono.MonoRuntimeType
+import com.jetbrains.rider.runtime.msNet.MsNetRuntimeType
 import org.jetbrains.concurrency.AsyncPromise
 import org.jetbrains.concurrency.Promise
+import org.jetbrains.concurrency.resolvedPromise
+import java.awt.event.FocusEvent
+import java.awt.event.FocusListener
+import java.util.*
+
+fun getFsiRunOptions(
+  project: Project,
+  configuration: FSharpScriptConfiguration? = null
+): Pair<DotNetExecutable, DotNetRuntime> {
+  val fsiHost = project.getComponent(FsiHost::class.java)
+  lateinit var sessionInfo: RdFsiSessionInfo
+  UIUtil.invokeLaterIfNeeded {
+    sessionInfo = fsiHost.rdFsiHost.requestNewFsiSessionInfo.sync(Unit)
+  }
```

Comment:
The `sessionInfo` variable won't be set in line 55 if you run this code from a non-EDT. So, this will break.

My current suggestion would be to annotate this with `@RequiresEdt` and remove `UIUtil.invokeLaterIfNeeded`.

But then, what trickery is this, is it a `sync()` call on EDT? It's not good, innit?

Is it possible to turn this thing into a suspending function?

Also, what's the reason you need EDT in the first place? Is it because of protocol threading? Then maybe annotate the model with `.async` and get rid of EDT requirement altogether?